In [ ]:
import gdown 
import os 
import pandas as pd
import numpy as np

### Load Dataset 

In [2]:
file_id = "1eztNdzA460wjWeqs-C-gF1L6tCI0a6ex"
url = f"https://drive.google.com/uc?id={file_id}"
output = "data/raw/flight_data_2024.csv"

gdown.download(url, output, quiet=False)

Downloading...
From (original): https://drive.google.com/uc?id=1eztNdzA460wjWeqs-C-gF1L6tCI0a6ex
From (redirected): https://drive.google.com/uc?id=1eztNdzA460wjWeqs-C-gF1L6tCI0a6ex&confirm=t&uuid=0d1215c5-e726-4ca9-9157-2d5442ae5cfe
To: /Users/aya/Documents/Flights_Delay/data/raw/flight_data_2024.csv
100%|██████████| 198M/198M [00:04<00:00, 42.6MB/s] 


'data/raw/flight_data_2024.csv'

### Read the dataset 

In [3]:
dataset = pd.read_csv(output)

### Review Dataset 

In [4]:
dataset.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 28 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   Unnamed: 0         1000000 non-null  int64  
 1   year               1000000 non-null  int64  
 2   month              1000000 non-null  int64  
 3   day_of_month       1000000 non-null  int64  
 4   day_of_week        1000000 non-null  int64  
 5   fl_date            1000000 non-null  str    
 6   distance           1000000 non-null  float64
 7   origin             1000000 non-null  str    
 8   origin_city_name   1000000 non-null  str    
 9   origin_state_nm    1000000 non-null  str    
 10  dest               1000000 non-null  str    
 11  dest_city_name     1000000 non-null  str    
 12  dest_state_nm      1000000 non-null  str    
 13  op_unique_carrier  1000000 non-null  str    
 14  op_carrier_fl_num  1000000 non-null  int64  
 15  crs_dep_time       1000000 non-null  int64  

In [5]:
dataset.head()

,Unnamed: 0,year,month,day_of_month,day_of_week,fl_date,distance,origin,origin_city_name,origin_state_nm,...,origin_lat,origin_lng,dest_lat,dest_lng,distance_zscore,distance_outlier,season,is_weekend,departure_hour,period_of_day
0,0,2024,1,6,6,2024-01-06,475.0,BNA,"Nashville, TN",Tennessee,...,36.124475,-86.678181,42.946932,-87.897064,-0.601835,0,winter,1,12,afternoon
1,1,2024,8,4,7,2024-08-04,2475.0,LAX,"Los Angeles, CA",California,...,33.942496,-118.408049,40.639928,-73.778692,2.746703,0,summer,1,11,morning
2,2,2024,1,12,5,2024-01-12,296.0,SJC,"San Jose, CA",California,...,37.362995,-121.928621,34.200694,-118.358667,-0.901530,0,winter,0,10,morning
3,3,2024,4,3,3,2024-04-03,2248.0,JFK,"New York, NY",New York,...,40.639928,-73.778692,36.080343,-115.152449,2.366644,0,spring,0,6,morning
4,4,2024,11,17,7,2024-11-17,1824.0,CLE,"Cleveland, OH",Ohio,...,41.409407,-81.854691,36.080343,-115.152449,1.656754,0,fall,1,12,afternoon


In [6]:
dataset.tail()

,Unnamed: 0,year,month,day_of_month,day_of_week,fl_date,distance,origin,origin_city_name,origin_state_nm,...,origin_lat,origin_lng,dest_lat,dest_lng,distance_zscore,distance_outlier,season,is_weekend,departure_hour,period_of_day
999995,999995,2024,11,10,7,2024-11-10,517.0,DEN,"Denver, CO",Colorado,...,39.861667,-104.673167,37.036378,-113.510303,-0.531516,0,fall,1,20,evening
999996,999996,2024,5,27,1,2024-05-27,610.0,LGA,"New York, NY",New York,...,40.777242,-73.872606,34.895671,-82.218859,-0.375809,0,spring,0,18,evening
999997,999997,2024,12,12,4,2024-12-12,386.0,SJC,"San Jose, CA",California,...,37.362995,-121.928621,36.080343,-115.152449,-0.750845,0,fall,0,20,evening
999998,999998,2024,5,20,1,2024-05-20,837.0,MSY,"New Orleans, LA",Louisiana,...,29.993272,-90.259028,41.976940,-87.908150,0.004250,0,spring,0,6,morning
999999,999999,2024,7,24,3,2024-07-24,1303.0,SAN,"San Diego, CA",California,...,32.733563,-117.189663,29.984435,-95.341442,0.784459,0,summer,0,13,afternoon


In [21]:
dataset.columns

Index(['Unnamed: 0', 'year', 'month', 'day_of_month', 'day_of_week', 'fl_date',
       'distance', 'origin', 'origin_city_name', 'origin_state_nm', 'dest',
       'dest_city_name', 'dest_state_nm', 'op_unique_carrier',
       'op_carrier_fl_num', 'crs_dep_time', 'crs_arr_time', 'delayed',
       'origin_lat', 'origin_lng', 'dest_lat', 'dest_lng', 'distance_zscore',
       'distance_outlier', 'season', 'is_weekend', 'departure_hour',
       'period_of_day'],
      dtype='str')

In [22]:
dataset = dataset.drop(columns=['Unnamed: 0'])

In [23]:
dataset.isna().sum()

year                 0
month                0
day_of_month         0
day_of_week          0
fl_date              0
distance             0
origin               0
origin_city_name     0
origin_state_nm      0
dest                 0
dest_city_name       0
dest_state_nm        0
op_unique_carrier    0
op_carrier_fl_num    0
crs_dep_time         0
crs_arr_time         0
delayed              0
origin_lat           0
origin_lng           0
dest_lat             0
dest_lng             0
distance_zscore      0
distance_outlier     0
season               0
is_weekend           0
departure_hour       0
period_of_day        0
dtype: int64

In [36]:
dataset[dataset['delayed'] == 1].count()

year                 208171
month                208171
day_of_month         208171
day_of_week          208171
fl_date              208171
distance             208171
origin               208171
origin_city_name     208171
origin_state_nm      208171
dest                 208171
dest_city_name       208171
dest_state_nm        208171
op_unique_carrier    208171
op_carrier_fl_num    208171
crs_dep_time         208171
crs_arr_time         208171
delayed              208171
origin_lat           208171
origin_lng           208171
dest_lat             208171
dest_lng             208171
distance_zscore      208171
distance_outlier     208171
season               208171
is_weekend           208171
departure_hour       208171
period_of_day        208171
dtype: int64

In [37]:
dataset[dataset['delayed'] == 0].count()

year                 791829
month                791829
day_of_month         791829
day_of_week          791829
fl_date              791829
distance             791829
origin               791829
origin_city_name     791829
origin_state_nm      791829
dest                 791829
dest_city_name       791829
dest_state_nm        791829
op_unique_carrier    791829
op_carrier_fl_num    791829
crs_dep_time         791829
crs_arr_time         791829
delayed              791829
origin_lat           791829
origin_lng           791829
dest_lat             791829
dest_lng             791829
distance_zscore      791829
distance_outlier     791829
season               791829
is_weekend           791829
departure_hour       791829
period_of_day        791829
dtype: int64

### Summarization , interpretation 

In [24]:
dataset.mode()

,year,month,day_of_month,day_of_week,fl_date,distance,origin,origin_city_name,origin_state_nm,dest,...,origin_lat,origin_lng,dest_lat,dest_lng,distance_zscore,distance_outlier,season,is_weekend,departure_hour,period_of_day
0,2024,7,15,1,2024-10-17,337.0,ATL,"Chicago, IL",Texas,ATL,...,33.6367,-84.427864,33.6367,-84.427864,-0.832885,0,fall,0,7,morning


### Interpretation of Mode 
The mode represents the most frequently occurring value in each column

### Categorical Data Analysis

## Cheack which season is the most season that may have delayed 

In [34]:
dataset[dataset['delayed'] == 1]["season"].value_counts()

season
summer    68168
spring    56217
fall      53836
winter    29950
Name: count, dtype: int64

## find the persantage of delay for each season 

In [49]:
dataset.groupby("season")["delayed"].mean().sort_values()

season
fall      0.160014
winter    0.200977
spring    0.221921
summer    0.260967
Name: delayed, dtype: float64

In [38]:
dataset[dataset['delayed'] == 1]["period_of_day"].value_counts()

period_of_day
afternoon    67666
evening      65055
morning      57363
night        18087
Name: count, dtype: int64

In [51]:
dataset.groupby("period_of_day")["delayed"].mean().sort_values()

period_of_day
morning      0.137376
afternoon    0.231360
night        0.268992
evening      0.292084
Name: delayed, dtype: float64

In [43]:
dataset[dataset['delayed'] == 1]["dest_city_name"].value_counts()

dest_city_name
Chicago, IL                       10515
Dallas/Fort Worth, TX             10470
Atlanta, GA                        8410
New York, NY                       8133
Denver, CO                         8077
                                  ...  
Moab, UT                              1
New Bern/Morehead/Beaufort, NC        1
Pago Pago, TT                         1
Dillingham, AK                        1
Morgantown, WV                        1
Name: count, Length: 338, dtype: int64

In [44]:
dataset[dataset['delayed'] == 1]["month"].value_counts()

month
7     26093
5     22924
6     21729
8     20346
1     18299
12    17625
3     17524
4     15769
9     12814
11    11927
2     11651
10    11470
Name: count, dtype: int64

In [46]:
dataset[dataset['delayed'] == 1]["day_of_month"].value_counts()

day_of_month
22    7700
15    7638
16    7378
19    7359
9     7310
21    7281
27    7159
8     7153
18    7150
28    7120
26    7067
29    6999
23    6976
17    6963
20    6907
12    6837
7     6789
24    6732
11    6710
5     6667
6     6600
14    6570
10    6503
2     6496
3     6451
25    6345
4     6268
30    5945
13    5898
1     5842
31    3358
Name: count, dtype: int64

## interpration Categorical Data Analysis
- we can see that summer is the season with the highest number of delays. Next, we have spring, followed closely by fall, which shows very similar results.
- we can see that afternoon is the period_of_day with the highest number of delays , then we have evening is very close to it  which it mean the most delayed cases happen on the afternoon and evening 
- we see that Chicago, IL  is the city that have the most delayed 

### Numerical Data Analysis

### Feature Correlation Analysis

In [41]:
dataset.corr(numeric_only=True)

,year,month,day_of_month,day_of_week,distance,op_carrier_fl_num,crs_dep_time,crs_arr_time,delayed,origin_lat,origin_lng,dest_lat,dest_lng,distance_zscore,distance_outlier,is_weekend,departure_hour
year,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
month,NaN,1.000000,0.003272,0.006569,-0.004969,0.031510,-0.003422,-0.008256,-0.038527,0.019580,-0.004784,0.016910,-0.006033,-0.004969,-0.002339,0.012381,-0.003318
day_of_month,NaN,0.003272,1.000000,0.012904,0.001442,0.003405,0.001653,-0.000545,0.009072,-0.000169,0.000835,-0.001001,-0.000486,0.001442,-0.001175,0.014172,0.001650
day_of_week,NaN,0.006569,0.012904,1.000000,0.008157,-0.005467,0.003731,-0.000665,0.027621,-0.008472,-0.002898,-0.008459,-0.002243,0.008157,0.001031,0.784152,0.003677
distance,NaN,-0.004969,0.001442,0.008157,1.000000,-0.372124,-0.008622,0.015483,0.016279,-0.047730,-0.115411,-0.048448,-0.116084,1.000000,0.344296,0.011612,-0.008828
op_carrier_fl_num,NaN,0.031510,0.003405,-0.005467,-0.372124,1.000000,0.027906,0.017634,-0.017495,0.093442,0.129872,0.098733,0.130769,-0.372124,-0.092020,-0.006005,0.027571
crs_dep_time,NaN,-0.003422,0.001653,0.003731,-0.008622,0.027906,1.000000,0.685170,0.161730,-0.031345,0.000153,0.031230,-0.007995,-0.008622,0.015623,-0.000825,0.999331
crs_arr_time,NaN,-0.008256,-0.000545,-0.000665,0.015483,0.017634,0.685170,1.000000,0.135774,-0.013375,0.015071,0.017287,0.004897,0.015483,-0.018303,-0.002098,0.684257
delayed,NaN,-0.038527,0.009072,0.027621,0.016279,-0.017495,0.161730,0.135774,1.000000,-0.014733,0.027353,-0.010078,0.015965,0.016279,-0.001716,0.011146,0.161462
origin_lat,NaN,0.019580,-0.000169,-0.008472,-0.047730,0.093442,-0.031345,-0.013375,-0.014733,1.000000,0.024407,0.229931,-0.016001,-0.047730,-0.045395,-0.011965,-0.031945


### Interpretation of Correlation Results

for this interpretaion i try to find the most strong and weak relationships 

- very strong relationship between departure_hour and crs_dep_time (0.9993), which is expected since departure_hour was directly bulid  from crs_dep_time. 
-  crs_dep_time and crs_arr_time share a moderate relationship (0.685), as flights departing later also tend to arrive later. 
- strong relationship between is_weekend and day_of_week (0.784), consistent with how the feature was constructed. 
-  moderate relationship between origin_lng and dest_lng (0.604), and origin_lat and dest_lat (0.230), reflecting that many flights operate within similar geographic corridors. 
- The distance and distance_zscore columns are perfectly correlated (1.000) since the z-score is derived directly from distance. 
- The distance_outlier and distance_zscore show a moderate positive relationship (0.344), as expected. 
- The year column produces NaN values across all pairs due to zero variance (all records are from 2024).


In [56]:
dataset.describe()

,year,month,day_of_month,day_of_week,distance,op_carrier_fl_num,crs_dep_time,crs_arr_time,delayed,origin_lat,origin_lng,dest_lat,dest_lng,distance_zscore,distance_outlier,is_weekend,departure_hour
count,1000000.0,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1.000000e+06,1000000.000000,1000000.000000,1000000.000000
mean,2024.0,6.597970,15.786584,3.983077,834.461610,2502.772380,1325.796484,1490.292545,0.208171,36.597999,-94.832342,36.596790,-94.843876,9.939072e-17,0.008571,0.276503,12.988599
std,0.0,3.396064,8.787083,2.010310,597.275568,1652.043715,492.424283,518.698428,0.406000,5.987153,18.546931,5.994319,18.559162,1.000000e+00,0.092182,0.447269,4.894416
min,2024.0,1.000000,1.000000,1.000000,31.000000,1.000000,1.000000,1.000000,0.000000,-14.331662,-176.642482,-14.331662,-176.642482,-1.345211e+00,0.000000,0.000000,0.000000
25%,2024.0,4.000000,8.000000,2.000000,399.000000,1150.000000,905.000000,1103.000000,0.000000,32.898639,-111.150260,32.898639,-111.150260,-7.290799e-01,0.000000,0.000000,9.000000
50%,2024.0,7.000000,16.000000,4.000000,680.000000,2224.000000,1320.000000,1515.000000,0.000000,36.894604,-87.908150,36.894604,-87.908150,-2.586103e-01,0.000000,0.000000,13.000000
75%,2024.0,10.000000,23.000000,6.000000,1069.000000,3717.000000,1735.000000,1925.000000,0.000000,40.777242,-80.951379,40.777242,-80.951379,3.926804e-01,0.000000,1.000000,17.000000
max,2024.0,12.000000,31.000000,7.000000,5095.000000,8819.000000,2359.000000,2359.000000,1.000000,71.284861,145.729986,71.284861,145.729986,7.133288e+00,1.000000,1.000000,23.000000


### Interpretation of describe Results

- All the dataset came form 2024 year and it's take the whole month and all days on the month and we can see that by the mean of the month (6.5) and day_of_month (15) and all the days on the week (3) 
- The average flight distance is approximately 834 miles, with a minimum of 31 miles and a maximum of 5,095 miles, indicating a wide range of short-haul to long-haul domestic flights. 
- Scheduled departure times (crs_dep_time) range from 0001 to 2359 with a mean around 1326 (early afternoon), while arrival times average around 1490.
- The target variable delayed has a mean of 0.208, confirming the class imbalance where approximately 20.8% of flights are delayed. 
- Latitude values for origin airports range from −14.3 to 71.3, and longitude from −176.6 to 145.7, reflecting the geographic diversity of U.S. domestic airports including Alaska, Hawaii, and U.S. territories.

### Feature Correlation Analysis how every numeric column relates to delayed

In [42]:
dataset.corr(numeric_only=True)["delayed"]

year                      NaN
month               -0.038527
day_of_month         0.009072
day_of_week          0.027621
distance             0.016279
op_carrier_fl_num   -0.017495
crs_dep_time         0.161730
crs_arr_time         0.135774
delayed              1.000000
origin_lat          -0.014733
origin_lng           0.027353
dest_lat            -0.010078
dest_lng             0.015965
distance_zscore      0.016279
distance_outlier    -0.001716
is_weekend           0.011146
departure_hour       0.161462
Name: delayed, dtype: float64

### Interpretation of Correlation Results relates to delayed 
We can see that most columns have almost no relationship with delayed. However, crs_dep_time, crs_arr_time, and departure_hour show a weak positive relationship with delayed, with correlation values of 0.161730, 0.135774, and 0.161462, respectively.

### Overall interpreation for the numerical Data Analysis

- The low correlation values confirm that flight delay prediction is a non-linear problem, which justifies the use of tree-based ensemble models such as Random Forest and XGBoost.